cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/co/cask/cdap/etl/planner/ControlDagTest.java

```diff
@@ -0,0 +1,318 @@
+    /*
+       trims down to:
+        n1 --> n2 --> n3 --> n4 --> n5 --> n6
+     */
+    Assert.assertEquals(5, cdag.trim());
+    expected = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n2", "n3"),
+        new Connection("n3", "n4"),
+        new Connection("n4", "n5"),
+        new Connection("n5", "n6")));
+    Assert.assertEquals(expected, cdag);
+  }
+
+  @Test
+  public void testFlattenNoOp() {
+    /*
+        n1 --> n2 --> n3
+     */
+    ControlDag cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n2", "n3")));
+    cdag.flatten();
+    ControlDag expected = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n2", "n3")));
+    Assert.assertEquals(expected, cdag);
+
+    /*
+             |--> n2 --|
+             |         |         |--> n6 --|
+        n1 --|--> n3 --|--> n5 --|         |--> n8
+             |         |         |--> n7 --|
+             |--> n4 --|
+     */
+    cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n3"),
+        new Connection("n1", "n4"),
+        new Connection("n2", "n5"),
+        new Connection("n3", "n5"),
+        new Connection("n4", "n5"),
+        new Connection("n5", "n6"),
+        new Connection("n5", "n7"),
+        new Connection("n6", "n8"),
+        new Connection("n7", "n8")));
+    cdag.flatten();
+    expected = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n3"),
+        new Connection("n1", "n4"),
+        new Connection("n2", "n5"),
+        new Connection("n3", "n5"),
+        new Connection("n4", "n5"),
+        new Connection("n5", "n6"),
+        new Connection("n5", "n7"),
+        new Connection("n6", "n8"),
+        new Connection("n7", "n8")));
+    Assert.assertEquals(expected, cdag);
+  }
+
+  @Test
+  public void testMultiSourceFlatten() {
+    /*
+        n1 --|
+             |--> n3
+        n2 --|
+     */
+    ControlDag cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n3"),
+        new Connection("n2", "n3")));
+    cdag.flatten();
+    ControlDag expected = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1.n2", "n1"),
+        new Connection("n1.n2", "n2"),
+        new Connection("n1", "n3"),
+        new Connection("n2", "n3")));
+    Assert.assertEquals(expected, cdag);
+  }
+
+  @Test
+  public void testFlatten() {
+    /*
+                      |--> n3
+            |--> n2 --|
+            |         |--> n4
+       n1 --|
+            |
+            |--> n5
+     */
+    ControlDag cdag = new ControlDag(
+      ImmutableSet.of(
+        new Connection("n1", "n2"),
+        new Connection("n1", "n5"),
+        new Connection("n2", "n3"),
+        new Connection("n2", "n4")));
+    cdag.flatten();
+    /*
```

Comment:
actually, scratch that. Flatten removes nested forks.  So this is an example where flattening does something less than ideal.